lms/djangoapps/courseware/masquerade.py

```diff
@@ -116,3 +174,146 @@ def get_masquerading_group_info(user, course_key):
     if not course_masquerade:
         return None, None
     return course_masquerade.group_id, course_masquerade.user_partition_id
+
+
+# Sentinel object to mark deleted objects in the session cache
+_deleted_sentinel = object()
+
+
+class MasqueradingFieldDataCache(object):
+    """
+    A wrapper for FieldDataCache used when masquerading as a different user.  Data is read from the
+    underlying cache, but written to the session, so the state of the actual user is not changed.
+    """
```

Comment:
@cpennington I chose to implement the FieldDataCache for two reasons.  First, the `set_score()` method directly modifies state in the database, which we don't want to do when masquerading as a different user.  Second, `field_data_cache` is created at the point where the masquerading is set up and then passed through many layers of function calls before it is actually used, so it was just a lot easier to implement it instead of a `FieldData`, since the creation and initialisation happened right in the place where the masquerading logic was anyway.

Regarding the first point, I just realized that overriding `set_score()` is not enough.  I rather have to make sure that the function `handle_grade_event()`, which calls `set_score()`, isn't called in the first place, since it makes several state-changing calls, so the first reason for my design decision is void.

The second point doesn't hold much water either, since I can simply check whether masquerading is active at the beginning of `get_module_system()`, and instantiate the correct `FieldData` depending on the result, so I don't have to do a lot of refactoring to make this work.

More specifically, I would implement a `MasqueradingKeyValueStore` that takes a generic `KeyValueStore` and a session as arguments, and implements the same logic as the `MasqueradingFieldDataCache`.  This `KeyValueStore` can be turned into a proper `FieldData` using the generic `KvsFieldData`, which is implemented in the XBlock code and is completely independent of the LMS machinery.

Would that be what you had in mind?